content/chapters/software-stack/lab/README.md

```diff
@@ -0,0 +1,952 @@
+* **fast development**: We want developers to be able to write code faster, using higher-level programming languages.
+
+The last two characteristics rely on two items:
+* **higher-level programming languages**: As discussed above, a compiler will take a higher-level program and transform it into binary code for different computing platforms, thus providing portability.
+  Also, it's easier to read (comprehend) and write (develop) source code in a higher-level programming language, thus providing fast development.
+* **software stacks**: A software stack is the layering of software such that each lower layer provides a set of features that the higher layer can directly use.
+  This means that there is no need for the higher layer to reimplement those features;
+  this provides fast development: focus on only the newer / required parts of software.
+
+  Also, each lower layer provides a generic interface to the higher layer.
+  These generic interfaces "hides" possible differences in the even lower layers.
+  This way, a software stack ensures portability across different other parts of software (and hardware as well).
+  For example, the standard C library, that we will present shortly, ensures portability across different operating systems.
+
+TODO: Diagram of generic software stack
+
+TODO: Quiz
+
+## Modern Software Stacks
+
+Most modern computing systems use a software stack such as the one in the figure below:
+
+TODO: modern software stack
+
+This modern software stack allows fast development and provides a rich set of applications to the user.
+
+The basic software component is the **operating system** (OS) (technically the operating system **kernel**).
+The OS provides the fundamental primitives to interact with hardware (read and write data) and to manage the running of applications (such as memory allocation, thread creation, scheduling).
+These primitives form the **system call API** or **system API**.
+An item in the system call API, i.e. the equivalent of a function call that triggers the execution of a functionality in the operating system, is a **system call**.
+
+The system call API is well defined, stable and complete: it exposes the entire functionality of the operating system and hardware.
+However, it is also minimalistic with respect to features and it provides a low-level (close to hardware) specification, making it cumbersome to use and **not portable**.
+
+Due to the downsides of the system call API, a basic library, the **standard C library** (also called **libc**), is built on top of it.
+The standard C library wraps each system call into an equivalent function call, following a portable calling convention.
+Because the system call API uses an OS-specific calling convention, the standard C library typically wraps each system call into an equivalent function call, following a portable calling convention.
+More than these wraps, the standard C library provides its own API that is typically portable.
```

Comment:
```suggestion
More than these wrappers, the standard C library provides its own API that is typically portable.
```
You don't state the reasons why libc adds functionality on top of the syscall wrappers clearly:
- syscalls don't provide enough functionality: `str*`, `prinft` vs `write` etc.
- using syscalls isn't optimal (hint at user space buffering, on which you expand later)
I think you can just list the disadvantageds of syscalls here and then go over them in greater detail later in the lab.